File: archive/t/typescript/dijkstra.ts
function printUsage(): void {
  console.log(
    "Usage: please provide three inputs: a serialized matrix, a source node and a destination node",
  );
}

function parseList(input: string | undefined): number[] | null {
  if (input === undefined) return null;
  if (input.trim() === "") return [];

  const parts = input.split(",").map((s) => s.trim());
  const nums: number[] = [];

  for (const p of parts) {
    if (!/^-?\d+$/.test(p)) return null;
    nums.push(Number(p));
  }

  return nums;
}

function buildMatrix(flat: number[]): number[][] | null {
  const n = Math.sqrt(flat.length);
  if ((n | 0) !== n) return null;

  const size = n | 0;
  const matrix: number[][] = new Array(size);

  for (let i = 0; i < size; i++) {
    const row: number[] = new Array(size);
    for (let j = 0; j < size; j++) {
      row[j] = flat[i * size + j];
    }
    matrix[i] = row;
  }

  return matrix;
}

function dijkstra(matrix: number[][], src: number, dst: number): number {
  const n = matrix.length;
  const dist = new Array(n).fill(Infinity);
  const visited = new Array(n).fill(false);

  dist[src] = 0;

  for (let i = 0; i < n; i++) {
    let u = -1;

    for (let j = 0; j < n; j++) {
      if (!visited[j] && (u === -1 || dist[j] < dist[u])) {
        u = j;
      }
    }

    if (u === -1 || dist[u] === Infinity) break;
    visited[u] = true;

    const row = matrix[u];
    const du = dist[u];

    for (let v = 0; v < n; v++) {
      const w = row[v];
      if (w > 0 && !visited[v]) {
        const alt = du + w;
        if (alt < dist[v]) dist[v] = alt;
      }
    }
  }

  return dist[dst] === Infinity ? -1 : dist[dst];
}

function main(): void {
  const [, , matrixStr, srcStr, dstStr] = process.argv;
  const flat = parseList(matrixStr);
  const srcArr = parseList(srcStr);
  const dstArr = parseList(dstStr);

  if (
    flat === null ||
    srcArr === null ||
    dstArr === null ||
    srcArr.length !== 1 ||
    dstArr.length !== 1
  ) {
    printUsage();
    return;
  }

  if (flat.length === 0) {
    printUsage();
    return;
  }

  const src = srcArr[0];
  const dst = dstArr[0];
  
  const matrix = buildMatrix(flat);
  if (!matrix) {
    printUsage();
    return;
  }

  const n = matrix.length;

  if (src < 0 || dst < 0 || src >= n || dst >= n) {
    printUsage();
    return;
  }

  for (const row of matrix) {
    for (const w of row) {
      if (w < 0) {
        printUsage();
        return;
      }
    }
  }

  const result = dijkstra(matrix, src, dst);
  if (result < 0) {
    printUsage();
    return;
  }

  console.log(result.toString());
}

main();

File: archive/r/ruby/dijkstra.rb
# frozen_string_literal: true

USAGE = "Usage: please provide three inputs: a serialized matrix, a source node and a destination node"

def usage!
  warn USAGE
  exit 1
end

def parse_matrix(str)
  str.split(",").map { Integer(it.strip) }
rescue ArgumentError, NoMethodError
  usage!
end

def build_matrix(flat)
  n = Integer(Math.sqrt(flat.size))
  usage! unless n * n == flat.size

  matrix = flat.each_slice(n).to_a

  usage! if matrix.any? { |row| row.any?(&:negative?) }

  matrix
end

def parse_index(str, n)
  usage! if str.nil? || str.strip.empty?

  i = Integer(str)
  usage! unless i.between?(0, n - 1)

  i
rescue ArgumentError
  usage!
end

def dijkstra(matrix, src, dst)
  n = matrix.size
  dist = Array.new(n, Float::INFINITY)
  visited = Array.new(n, false)

  dist[src] = 0

  n.times do
    u = nil
    best = Float::INFINITY

    n.times do |i|
      next if visited[i]
      next if dist[i] >= best

      best = dist[i]
      u = i
    end

    break unless u

    visited[u] = true

    row = matrix[u]
    base = dist[u]

    row.each_with_index do |w, v|
      next if w <= 0 || visited[v]

      alt = base + w
      dist[v] = alt if alt < dist[v]
    end
  end

  dist[dst]
end

matrix_str, src_str, dst_str = ARGV
usage! if matrix_str.nil? || src_str.nil? || dst_str.nil?
usage! if matrix_str.strip.empty? || src_str.strip.empty? || dst_str.strip.empty?

matrix = build_matrix(parse_matrix(matrix_str))
n = matrix.size

src = parse_index(src_str, n)
dst = parse_index(dst_str, n)

dist = dijkstra(matrix, src, dst)

usage! unless dist.finite?

puts dist.to_i
